doc/TAM/SAI-Proposal-TAM-stream-telemetry.md

```diff
@@ -0,0 +1,967 @@
+| TAM_collector            | Global                                |
+| TAM                      | Global                                |
+| TAM_telemetry            | per profile                           |
+| TAM_telemetry_type       | per profile / per type of one profile |
+| TAM_report               | per profile / per type of one profile |
+| TAM_counter_subscription | per stats of object                   |
+
+#### TAM Telemetry state machine
+
+``` mermaid
+
+---
+title: TAM Telemetry state machine
+---
+stateDiagram-v2
+    direction LR
+
+    state "Stop TAM Stream" as stop
+    state "Start TAM Stream" as start
+    state "Commit configure" as commit
+    state "Config ready" as ready
+    state join_commit <<join>>
+
+    [*] --> stop: Object create
+    stop --> start: NOS set
+    stop --> join_commit: NOS set
+
+    start --> join_commit: NOS set (The record stream will not stop and original templates will still be used)
+    join_commit --> commit: Vendor start preparing templates based on bounded telemetry types
+    start --> stop: NOS set
+
+    commit --> ready: Vendor set (callback via notification)
+
+    ready --> start: NOS set (New templates will be applied to generate the new record streaming)
+    ready --> stop: NOS set
+
+    stop --> [*]: Object remove
+
+```
+
+#### Initialize TAM cache count for Switch
+
+``` c++
+/**
+ * @brief Attribute Id in sai_set_switch_attribute() and
+ * sai_get_switch_attribute() calls.
+ */
+typedef enum _sai_switch_attr_t
+{
+    // ...
+
+    /**
+     * @brief Tam telemetry buffer size
+     *
+     * If the collector isn't ready to receive reports, this value indicates how many
+     * bytes of reports that can be cached.
+     * 0 means no cache which is the default behavior.
+     * If the cache is full, new incoming data will be dropped.
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_ONLY
+     * @default 0
+     */
+    SAI_SWITCH_ATTR_TAM_BUFFER_SIZE,
+
+    /**
+     * @brief Tam telemetry reporting chunk size
+     *
+     * Defines the maximum number of bytes in a single report.
+     * The total number of bytes in a report should be as close as possible to this value.
+     * We can increase this value to reduce the number of sys calls.
+     * If the type of message is IPFIX, this value should not be less than 65535.
+     * Because we don't expect the IPFIX record to be fragmented.
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 65535
+     */
+    SAI_SWITCH_ATTR_TAM_CHUNK_SIZE,
+
+    /**
+     * @brief Set TAM telemetry state change event notification callback function passed to the adapter.
+     *
+     * Use sai_tam_telemetry_state_change_notification_fn as notification function.
+     *
+     * @type sai_pointer_t sai_tam_telemetry_state_change_notification_fn
+     * @flags CREATE_AND_SET
+     * @default NULL
+     */
+    SAI_SWITCH_ATTR_TAM_TELEMETRY_STATE_CHANGE_NOTIFY,

```

Comment:
will this one be better to be put on TAM_TEL_TYPE?

in this way, if we change the state of 1 tel_type (say, queue), we won't impact the state of another type (say, port), in the non-mixed case.

Reply:
Done